File: native-views/react-native-native-list/android/src/main/java/com/onekey/nativelist/HybridNativeList.kt
package com.margelo.nitro.nativelist

import android.os.Looper
import android.view.View
import com.facebook.proguard.annotations.DoNotStrip
import com.facebook.react.uimanager.ThemedReactContext

@DoNotStrip
class HybridNativeList(context: ThemedReactContext) : HybridNativeListSpec() {
  private val hostView = NativeListView(context)

  override val view: View = hostView

  override var snapshotJson: String = ""
    get() = field
    set(value) {
      if (field == value) return
      field = value
      dispatchToUi { hostView.applySnapshot(value) }
    }

  override var onRowAction: ((payloadJson: String) -> Unit)? = null
    set(value) { field = value; hostView.onRowAction = value }
  override var onSelectionDelta: ((payloadJson: String) -> Unit)? = null
    set(value) { field = value; hostView.onSelectionDelta = value }
  override var onReorder: ((payloadJson: String) -> Unit)? = null
    set(value) { field = value; hostView.onReorder = value }
  override var onEndReached: ((payloadJson: String) -> Unit)? = null
    set(value) { field = value; hostView.onEndReached = value }
  override var onVisibleRangeChanged: ((payloadJson: String) -> Unit)? = null
    set(value) { field = value; hostView.onVisibleRangeChanged = value }

  override fun applySnapshot(snapshotJson: String) {
    dispatchToUi { hostView.applySnapshot(snapshotJson) }
  }

  override fun applyPatches(patchesJson: String) {
    dispatchToUi { hostView.applyPatches(patchesJson) }
  }

  override fun reconcileSelection(selectedKeysJson: String) {
    dispatchToUi { hostView.reconcileSelection(selectedKeysJson) }
  }

  override fun scrollToKey(
    key: String,
    animated: Boolean,
    alignment: NativeListScrollAlignment,
    viewPosition: Double,
    viewOffset: Double,
  ) {
    dispatchToUi {
      hostView.scrollToKey(
        key,
        animated,
        alignment.stringValue(),
        viewPosition,
        viewOffset,
      )
    }
  }

  override fun scrollToIndex(
    index: Double,
    animated: Boolean,
    alignment: NativeListScrollAlignment,
    viewPosition: Double,
    viewOffset: Double,
  ) {
    if (!index.isFinite() || index < 0 || index % 1.0 != 0.0 || index > Int.MAX_VALUE) return
    dispatchToUi {
      hostView.scrollToIndex(
        index.toInt(),
        animated,
        alignment.stringValue(),
        viewPosition,
        viewOffset,
      )
    }
  }

  override fun scrollToOffset(offset: Double, animated: Boolean) {
    if (!offset.isFinite() || offset < 0) return
    dispatchToUi { hostView.scrollToOffset(offset, animated) }
  }

  override fun scrollToEnd(animated: Boolean) {
    dispatchToUi { hostView.scrollToEnd(animated) }
  }

  override fun setRefreshing(refreshing: Boolean) {
    dispatchToUi { hostView.setRefreshing(refreshing) }
  }

  override fun onDropView() {
    hostView.dispose()
    super.onDropView()
  }

  override fun dispose() {
    hostView.dispose()
    super.dispose()
  }

  /**
   * Nitro prop setters can run off the Android UI thread. Apply synchronously
   * during a UI commit; only off-main calls need to be posted to this host.
   */
  private fun dispatchToUi(action: () -> Unit) {
    if (Looper.myLooper() == Looper.getMainLooper()) action() else hostView.post(action)
  }

  private fun NativeListScrollAlignment.stringValue(): String = when (this) {
    NativeListScrollAlignment.START -> "start"
    NativeListScrollAlignment.CENTER -> "center"
    NativeListScrollAlignment.END -> "end"
    NativeListScrollAlignment.NEAREST -> "nearest"
  }
}

File: native-views/react-native-native-list/ios/HybridNativeList.swift
import Foundation
import UIKit

final class HybridNativeList: HybridNativeListSpec {
  private let hostView = NativeListView(frame: .zero)

  var view: UIView { hostView }

  var snapshotJson: String = "" {
    didSet {
      guard snapshotJson != oldValue else { return }
      runOnMain { [weak self] in
        guard let self else { return }
        self.hostView.applySnapshotJson(self.snapshotJson)
      }
    }
  }

  var onRowAction: ((_ payloadJson: String) -> Void)?
  var onSelectionDelta: ((_ payloadJson: String) -> Void)?
  var onReorder: ((_ payloadJson: String) -> Void)?
  var onEndReached: ((_ payloadJson: String) -> Void)?
  var onVisibleRangeChanged: ((_ payloadJson: String) -> Void)?

  override init() {
    super.init()
    hostView.onRowAction = { [weak self] in self?.onRowAction?($0) }
    hostView.onSelectionDelta = { [weak self] in self?.onSelectionDelta?($0) }
    hostView.onReorder = { [weak self] in self?.onReorder?($0) }
    hostView.onEndReached = { [weak self] in self?.onEndReached?($0) }
    hostView.onVisibleRangeChanged = { [weak self] in self?.onVisibleRangeChanged?($0) }
  }

  func applySnapshot(snapshotJson: String) throws {
    runOnMain { [weak self] in self?.hostView.applySnapshotJson(snapshotJson) }
  }

  func applyPatches(patchesJson: String) throws {
    runOnMain { [weak self] in self?.hostView.applyPatchesJson(patchesJson) }
  }

  func reconcileSelection(selectedKeysJson: String) throws {
    runOnMain { [weak self] in self?.hostView.reconcileSelectionJson(selectedKeysJson) }
  }

  func scrollToKey(
    key: String,
    animated: Bool,
    alignment: NativeListScrollAlignment,
    viewPosition: Double,
    viewOffset: Double
  ) throws {
    runOnMain { [weak self] in
      self?.hostView.scrollToKey(
        key,
        animated: animated,
        alignment: alignment.stringValue,
        viewPosition: viewPosition,
        viewOffset: viewOffset
      )
    }
  }

  func scrollToIndex(
    index: Double,
    animated: Bool,
    alignment: NativeListScrollAlignment,
    viewPosition: Double,
    viewOffset: Double
  ) throws {
    guard index.isFinite,
          index >= 0,
          index.rounded(.towardZero) == index,
          index <= Double(Int.max) else { return }
    runOnMain { [weak self] in
      self?.hostView.scrollToIndex(
        Int(index),
        animated: animated,
        alignment: alignment.stringValue,
        viewPosition: viewPosition,
        viewOffset: viewOffset
      )
    }
  }

  func scrollToOffset(offset: Double, animated: Bool) throws {
    guard offset.isFinite, offset >= 0 else { return }
    runOnMain { [weak self] in
      self?.hostView.scrollToOffset(offset, animated: animated)
    }
  }

  func scrollToEnd(animated: Bool) throws {
    runOnMain { [weak self] in self?.hostView.scrollToEnd(animated: animated) }
  }

  func setRefreshing(refreshing: Bool) throws {
    runOnMain { [weak self] in self?.hostView.setRefreshing(refreshing) }
  }

  private func runOnMain(_ work: @escaping () -> Void) {
    if Thread.isMainThread {
      work()
    } else {
      DispatchQueue.main.async(execute: work)
    }
  }
}
